[BEAM-8603] Add Python SqlTransform

TheNeuralBit opened the pull request:
This PR adds a cross-language `SqlTransform` stub to the python SDK. For expansion, it relies on a new SQL-only expansion service (`:sdks:java:extensions:sql:expansion-service`) added to Java. It also adds a `wordcount_xlang_sql` example script, and a suite of tests, `sql_test.py`, which can be run continuously with the new `:validatesCrossLanguageRunnerPythonUsingSql` gradle target.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

TheNeuralBit commented:
CC: @robinyqiu

TheNeuralBit commented on sdks/java/extensions/sql/build.gradle:
```diff
@@ -24,6 +24,7 @@ plugins {
 }
 applyJavaNature(
   automaticModuleName: 'org.apache.beam.sdk.extensions.sql',
+  shadowClosure: {},
```
@lukecwik I added this so I can build a single jar to pass to Java workers for the xlang transform (see the `--experiment='jar_packages=./sdks/java/extensions/sql/build/libs/beam-sdks-java-extensions-sql-2.18.0-SNAPSHOT.jar' \` line in my "how to run" in the description), but it looks like it's causing `:sdks:java:extensions:sql:validateShadedJarDoesntLeakNonProjectClasses` to fail spectacularly. Is there some other way I can do this?

lukecwik replied:
I don't think we want to do this since normal SQL users don't want to be using an artifact meant for the expansion service. It seems as though we should have the expansion service gradle project build one uber artifact containing all the xlang transforms we want to vend.

We could have multiple shadowJar tasks and produce multiple artifacts along this line but this seems overly complex initially.

@chamikaramj Is there an initial strategy we are following for how we want to build xlang expansion service jars?

chamikaramj replied:
We need two jars.

(1) Jar for expansion service that contains the transform classes that need to be expanded (this can be multiple jars as well, IO transforms just have to be in the class path to be picked up by the AutoService).
(2) Jar to be passed to the Java worker. For me, building a shadow Jar of java-harness (:sdks:java:harness:shadowJar) worked.

It probably makes sens to move expansion service to it's own Gradle module and support building a shadow jar with all in-built cross-language transforms in Beam and release that.

robinyqiu commented:
Thank you Brian for putting this PR together and writing up the instructions. That's very helpful! I run the demo in this PR at Apache Beam Zurich meetup and it went very well!

TheNeuralBit commented:
Thanks stalebot. I'd still like to look at this.

TheNeuralBit commented:
I force pushed some significant changes to this PR. Now rather than adding SQL as a dependency in the test expansion service, this adds a new `:sdks:java:extensions:sql:expansion-service` artifact that produces a shaded jar with `SqlTransform` registered. The `wordcount_xlang_sql` script is meant to be called with this jar, which it will execute and use as the expansion service, and _also_ send to workers via the `jar_packages` experiment.

The PR also includes the changes from #11008, which should be merged first.

@chamikaramj can you take another look now?

TheNeuralBit commented:
Run Java PreCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
R: @robertwb could you review this?

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

robertwb commented on sdks/python/apache_beam/transforms/sql.py:
```diff
@@ -0,0 +1,44 @@
+# limitations under the License.
+#
+
+"""Transforms that use the SQL Expansion service."""
```
The fact that it's using an SQL Expansion service is an implementation detail. Instead, I would document this as if they were natively written, e.g. "Transforms that allow one to perform SQL operations on Beam PCollections" with at least a cursory note about schemas, and perhaps an example (or at least pointer to the test as one), and then just a note that the actual implementation is shared with java.

mxm replied:
+1 The goal should be that these things work seamlessly and the docs should reflect that.

robertwb commented on sdks/python/apache_beam/transforms/sql.py:
```diff
@@ -0,0 +1,44 @@
+    'SqlTransformSchema', [('query', unicode)])
+
+
+class SqlTransform(ExternalTransform):
```
This class should have a docstring, even if it's just a reference to the Beam documentation on SQL.

mxm replied:
+1

robertwb replied:
+1, very nice.

robertwb commented on sdks/python/apache_beam/transforms/sql.py:
```diff
@@ -0,0 +1,44 @@
+        self.URN,
+        NamedTupleBasedPayloadBuilder(SqlTransformSchema(query=query)),
+        BeamJarExpansionService(
+            ':sdks:java:extensions:sql:expansion-service:shadowJar'))
```
I've been thinking on how to best test these things, and possibly it could make sense to have a mocking infrastructure (or even pipeline options) where one could provide a mapping from these hard-coded targets to explicit paths to jars and/or URLs to pre-started services. 

(Likely in a follow-up PR.)

mxm replied:
It's also odd that this will attempt to download when the build order is not correct (jars are missing). Perhaps we need a TestExpansionService which can be augmented better like you suggested.

robertwb commented on sdks/python/apache_beam/transforms/sql_test.py:
```diff
@@ -0,0 +1,109 @@
+@unittest.skipIf(
+    TestPipeline().get_pipeline_options().view_as(StandardOptions).runner is
+    None,
+    "Must be run with a runner that supports cross-language transforms")
```
The default (Python) runner does support cross-language transforms.

TheNeuralBit replied:
Oh ok I didn't realize that.

Really I just needed a way to prevent this test from running in the Python PreCommit, since the SQL expansion service isn't built in that context. The other xlang test suite handles that by checking if the `EXPANSION_PORT` env var is set.

TheNeuralBit replied:
Ah actually I also ran into an issue when running this test with the default runner, that I'm having a hard time making sense of:

```
E       ValueError: Missing requirement declaration: {'beam:requirement:pardo:splittable_dofn:v1'}                     
                                                                                                                       
sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:651: ValueError
```

It looks like it's indicating my pipeline should contain a `splittable_dofn` declaration but doesn't - but I'm not clear on why it needs that declaration (I don't think anything in the pipeline needs to be splittable), or how to add it.

mxm replied:
Is this still an issue? Do you have the complete stack trace?

robertwb commented on sdks/python/apache_beam/transforms/sql_test.py:
```diff
@@ -0,0 +1,109 @@
+
+  Alternatively, you may be able to iterate faster if you run the tests directly
+  using a runner like `FlinkRunner`, which starts its own job server, but you'll
+  need to spin up a local flink cluster:
```
You shouldn't need to spun up a flink cluster, just using `--runner=FlinkRunner` should be sufficient. (As mentioned, you don't even need to pass a runner, as the default runner works as well.)

mxm replied:
+1

robertwb commented on sdks/python/apache_beam/transforms/sql_test.py:
```diff
@@ -0,0 +1,109 @@
+        ":sdks:java:extensions:sql:expansion-service:shadowJar")
+    test_pipeline = TestPipeline()
+    test_pipeline.get_pipeline_options().view_as(DebugOptions).experiments = [
+        'jar_packages=' + path_to_jar
```
@ihji What JIRA should we reference to obviate the need to do this?

ihji replied:
We can remove `jar_packages` flag when BEAM-9238 is done.

robertwb commented on sdks/python/apache_beam/transforms/sql_test.py:
```diff
@@ -0,0 +1,109 @@
+    test_pipeline.get_pipeline_options().view_as(DebugOptions).experiments = [
+        'jar_packages=' + path_to_jar
+    ]
+    test_pipeline.not_use_test_runner_api = True
```
Is this still needed? If so, maybe it won't be after @chamikaramj 's fixes? (JIRA?)

robertwb commented on sdks/python/apache_beam/transforms/sql_test.py:
```diff
@@ -0,0 +1,109 @@
+          | beam.Create([SimpleRow(1, "foo", 3.14), SimpleRow(2, "bar", 1.414)])
+          | SqlTransform("SELECT * FROM PCOLLECTION WHERE `str` = 'bar'"))
+      assert_that(out, equal_to([(2, "bar", 1.414)]))
+
```
It'd be great to have at least one grouping/combining test.

robertwb commented on sdks/python/apache_beam/examples/wordcount_xlang_sql.py:
```diff
@@ -0,0 +1,136 @@
+coders.registry.register_coder(MyRow, coders.RowCoder)
+
+
+class WordExtractingDoFn(beam.DoFn):
```
Nuke this class and just do `FlatMap(lambda line: re.findall(r'[\w\']+', line))` or, since this is not the focus of this example, even simply `FlatMap(str.split)` and avoid the save main session ugliness.

mxm replied:
+1

robertwb commented on sdks/python/apache_beam/examples/wordcount_xlang_sql.py:
```diff
@@ -0,0 +1,136 @@
+      | 'sql!!' >> SqlTransform(
+          """
+                   SELECT
+                     word as key,
```
Does the `as key` add value? (Similarly above.)

robertwb commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/SqlTransform.java:
```diff
@@ -271,6 +275,19 @@ static Builder builder() {
+    @Override
+    public PTransform<PInput, PCollection<Row>> buildExternal(
+        External.Configuration configuration) {
+      return builder()
```
@chamikaramj @chadrik It'd be nice to reduce this boilerplate to almost nothing, given we're already using the builder pattern here, or a minimal mapping of names (e.g. query -> queryString).

mxm replied:
I agree that a simple mapping would be nice. The current interface could still be useful for the more advanced external configurations. Does not seem to be of very high priority though.

TheNeuralBit commented:
> I reviewed everything but the groovy files, which I would like another set of eyes on.

R: @ihji could you take a look at the groovy changes?

ihji commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -1779,6 +1779,33 @@ class BeamModulePlugin implements Plugin<Project> {
+        "suite": "xlangSqlValidateRunner"
+      ]
+      def cmdArgs = project.project(':sdks:python').mapToArgString(beamPythonTestPipelineOptions)
+      def pythonSqlTask = project.tasks.create(name: config.name+"PythonUsingSql", type: Exec) {
```
You will run the same test twice if you put the task here without using `sdk` and `port`. It will run twice against Python expansion-service and Java expansion-service but `pythonSqlTask` uses neither of them.

I would instead add `:sdks:java:extensions:sql` dependency to existing test expansion service and utilize environment variables `EXPANSION_JAR` and `EXPANSION_PORT` like `pythonTask` above. Since external sql transform is only available on Java expansion service, you should handle the exception like https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/external/xlang_parquetio_test.py#L71

ihji replied:
If it's outside of the each block then it should be fine. The advantage of using existing testing expansion service here is you don't need to start and stop another expansion service inside the test code. The expansion service starts and finishes at the beginning and the end of the whole test suite. You can just pass the jar name and the port number by environment variables. Still there's no harm in launching your own expansion service though.

mxm commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/SqlTransform.java:
```diff
@@ -296,4 +313,24 @@ static UdafDefinition of(String udafName, Combine.CombineFn combineFn) {
+    }
+
+    public static class Configuration {
+      String query;
```
Do we need to add the dialect used here?

TheNeuralBit commented:
Run Python PreCommit

TheNeuralBit commented:
Run Python PreCommit

TheNeuralBit commented:
Run CommunityMetrics PreCommit

TheNeuralBit commented:
Run Python2_PVR_Flink PreCommit

TheNeuralBit commented:
Run CommunityMetrics PreCommit

TheNeuralBit commented:
It looks like PreCommit failures are just flakes or persistent failures at this point.

I think I've addressed all of the PR comments, I just need an answer on two things:
- @robertwb: Is it acceptable to skip sql_test for the default runner for now, since it doesn't support  java_artifacts?
- @ihji: Are the groovy changes ok given that they're outside of the `.each` block?

TheNeuralBit commented:
cc: @Ardagan

TheNeuralBit commented:
ping @robertwb @ihji

TheNeuralBit commented:
Run Java PreCommit

TheNeuralBit commented:
Run Python PreCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Merged past https://github.com/apache/beam/pull/11428 to hopefully resolve precommit issues.

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run Python PreCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommi

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Flink PostCommit

TheNeuralBit commented:
Run XVR_Spark PostCommit

TheNeuralBit commented:
Run Java PreCommit

aaltay commented:
Nice! Thank you!

pabloem commented:
yoohooo
